### PR TITLE
feature/Bias mitigation enhancement

### PR DESCRIPTION
**Pull request by mingkang111**
# Description
This PR aims to incorporate the bias mitigation enhancement from BB into guardian-ai.

# New Dependency
* `fairlearn==0.10.0`

# Change log
* Several enhancements on bias mitigation and fairness metrics.
* Relaxed version restriction for `numpy`, `pandas`, `scikit-learn==1.3.2` and `scipy==1.10.0`.
* Added test cases.
* Fixed doc.

### CODE REVIEW

File: guardian_ai/fairness/metrics/utils.py [hunk 1]
```diff
@@ -457,3 +481,118 @@ def _place_space_before_capital_letters(input_str):
     capital_letter_words = [word.strip() for word in capital_letter_words]
 
     return " ".join(capital_letter_words)
+
+
+# TODO: Move everything below here to the metrics module
```
Comment (YashaPushak) [inline guardian_ai/fairness/metrics/utils.py:486]: TODO appears to be stale -- you can delete it.

File: guardian_ai/fairness/metrics/utils.py [hunk 2]
```diff
@@ -457,3 +481,118 @@ def _place_space_before_capital_letters(input_str):
...
+
+        return FP / (TP + FP)
+    elif rate == "FOR":
+        # False ommission rate
+        FN = _FN(y_true, y_pred)
+        TN = _TN(y_true, y_pred)
+
+        return FN / (FN + TN)
+    else:
+        raise ValueError(f"Undefined rate {rate}")
```
Comment (YashaPushak) [inline guardian_ai/fairness/metrics/utils.py:559]: Minor: I believe that there are Guardian AI specific value errors, right?

File: guardian_ai/fairness/utils/util.py [hunk 3]
```diff
@@ -7,71 +7,71 @@
...
-        "neg_mean_abs_error",
-    ],
-    # metrics starting with 'unsupervised' do not require contamination factor to be provided.
-    "unsupervised": [
-        # "unsupervised_n-1_experts",
-        "unsupervised_unify95",
-        "unsupervised_unify95_log_loss",
-    ],
-}
+_supported_score_metric: Dict[str, List[str]] = (
```
Comment (YashaPushak) [inline guardian_ai/fairness/utils/util.py:10]: The fairness module only works for binary classification. I'm not sure if this dict actually gets used to access the binary metrics, but at least the rest of these are not strictly required to be here, as the other tasks aren't supported.

File: guardian_ai/fairness/bias_mitigation/sklearn.py [hunk 4]
```diff
@@ -267,9 +287,90 @@ def __init__(
         self._unique_group_names_: Optional[list] = None
         self._multiplier_names_: Optional[List[str]] = None
         self._admissible_trials_mask_: Optional[pd.DataFrame] = None
+        self._third_objective = third_objective
+        self._third_objective_name = "Outcome Regression"
```
Comment (YashaPushak) [inline guardian_ai/fairness/bias_mitigation/sklearn.py:291]: We ended up renaming this to the more commonly used term "Levelling Down"

File: guardian_ai/fairness/bias_mitigation/sklearn.py [hunk 5]
```diff
@@ -1294,6 +1494,171 @@ def selected_multipliers_(self):  # noqa D102
...
+            predictions = self._probas_predicted.argmax(-1)
+            rates_by_group = []
+            for group_name in self._unique_group_names_:
+                mask = group_masks[group_name]
+                rates_by_group.append(rate_scorer(y[mask], predictions[mask]))
+
+            # print(rates_by_group)
+            # Find the group with the best outcome rate
+            target_rate = best(rates_by_group)
+            # print(f'OR target rate: {target_rate}')
```
Comment (YashaPushak) [inline guardian_ai/fairness/bias_mitigation/sklearn.py:1586]: Probably worth deleting all of the print statements that are commented out that I was using for debugging?

File: guardian_ai/fairness/bias_mitigation/sklearn.py [hunk 6]
```diff
@@ -1154,13 +1351,16 @@ def show_tradeoff(self, hide_inadmissible: bool = False):
...
                 y=df[self.accuracy_metric_name],
+                customdata=df[self._third_objective_name],
                 text=df["index"],
                 line_shape="vh" if self._higher_fairness_is_better else "hv",
                 mode="markers+lines",
                 hovertemplate=f"{self.fairness_metric_name}"
                 + ": %{x:.4f}"
                 + f"<br>{self.accuracy_metric_name}"
                 + ": %{y:.4f}</br>"
+                + f"<br>TPR Difference"
```
Comment (YashaPushak) [inline guardian_ai/fairness/bias_mitigation/sklearn.py:1362]: "TPR Difference" should be renamed to "Levelling Down"

File: tests/unitary/test_fairness_bias_mitigation.py [hunk 7]
```diff
@@ -225,10 +225,14 @@ def test_group_ranges(sensitive_dataset_and_model):
         accuracy_metric=AN_ACCURACY_METRIC,
         random_seed=RANDOM_SEED,
     )
-
```
Comment (YashaPushak) [inline tests/unitary/test_fairness_bias_mitigation.py:228]: on line 25, there is a constant defined "A_FAIRNESS_METRIC". It's currently set to "equalized_odds", but this PR introduces changes to the bias mitigation algorithm that only apply to some metrics and equalized_odds isn't one of them. It's probably best to change the constant to be a list that includes "equalized_odds" and "TPR" and then to run any test case that uses the constant twice.

File: guardian_ai/fairness/metrics/model.py [hunk 8]
```diff
@@ -36,6 +36,16 @@
...
+_valid_regression_metrics = [
+    "TPR",
+    "statistical_parity",
+    "FPR",
+    "FNR",
+    "FOR",
+    "FDR",
+    "error_rate",
+]
+
```
Comment (YashaPushak) [inline guardian_ai/fairness/metrics/model.py:48]: I reviewed the changes to this file that haven't been ported yet. In the other repo, the function `_model_metric` is called `_all_group_pairs_model_metric`, but `_model_metric` is a better name and should be preferred. However, the rest of the changes that happen inside of that function should be ported here.

The function `_outcome_rate_metric` that is in the other repo appears to be stale and unused, so does not need to be ported.

File: guardian_ai/fairness/bias_mitigation/sklearn.py [hunk 9]
```diff
@@ -1154,13 +1351,16 @@ def show_tradeoff(self, hide_inadmissible: bool = False):
             go.Scatter(
                 x=df[self.fairness_metric_name],
                 y=df[self.accuracy_metric_name],
+                customdata=df[self._third_objective_name],
```
Comment (YashaPushak) [inline guardian_ai/fairness/bias_mitigation/sklearn.py:1354]: Can you also update this plot to show `df[self._third_objective_name]` as the colour of each point? The colour should be on a red-blue scale, where blue values are 0 and red values are farther from 0. 

The name of the colour in the legend should be "Levelling Down"

The lines plotted between the points in the scatter plot will also need to be removed.

Comment (YashaPushak) [reply guardian_ai/fairness/bias_mitigation/sklearn.py:1354]: Note that this function probably also needs to be updated to have two versions of the plot -- if the fairness metric name is not in `_valid_regression_metrics`, then the original version of the plot should be displayed (don't use any of the changes from this PR); otherwise, the new version should be displayed that includes the changes that are being ported here + the ones that I described above.

**Review by YashaPushak — COMMENTED** [review]
Thanks @mingkang111! The changes that you have ported so far look good. I left a few minor comments and a couple larger ones related to the model metric file that is missing some parts to be ported and the visualization in show_tradeoff.

File: guardian_ai/fairness/utils/util.py (hunk 3 above)
Comment (mingkang111) [reply guardian_ai/fairness/utils/util.py]: Removed the other keys.

File: guardian_ai/fairness/bias_mitigation/sklearn.py (hunk 4 above)
Comment (mingkang111) [reply guardian_ai/fairness/bias_mitigation/sklearn.py]: Changed the constant to Levelling Down.

File: guardian_ai/fairness/bias_mitigation/sklearn.py (hunk 5 above)
Comment (mingkang111) [reply guardian_ai/fairness/bias_mitigation/sklearn.py]: Removed.

File: guardian_ai/fairness/bias_mitigation/sklearn.py (hunk 9 above)
Comment (mingkang111) [reply guardian_ai/fairness/bias_mitigation/sklearn.py:1358]: I have updated the `show_tradeoff` as requested.  The following is the plot generated by test. 
Screenshot 2024-04-09 at 17 20 00

File: guardian_ai/fairness/metrics/model.py (hunk 8 above)
Comment (mingkang111) [reply guardian_ai/fairness/metrics/model.py:52]: By incorporating the the rest of the changes that happen inside of that `_all_group_pairs_model_metric`, one of the test `test_calibration_model_scorers_two_classes` start failing.

File: tests/unitary/test_fairness_bias_mitigation.py (hunk 7 above)
Comment (mingkang111) [reply tests/unitary/test_fairness_bias_mitigation.py:228]: Test updated: 
Screenshot 2024-04-09 at 17 46 08
Screenshot 2024-04-09 at 17 50 31

File: guardian_ai/fairness/bias_mitigation/sklearn.py (hunk 9 above)
Comment (YashaPushak) [reply guardian_ai/fairness/bias_mitigation/sklearn.py:1358]: Thanks! The levelling down colour bar should be capped at 0. Depending on whether or not the numbers have been multiplied by -1, it should either be 0 and up or 0 and down. Either way, 0 should be blue and away from 0 should be red.

File: guardian_ai/fairness/metrics/model.py (hunk 8 above)
Comment (mingkang111) [reply guardian_ai/fairness/metrics/model.py:52]: Fixed.

File: guardian_ai/fairness/bias_mitigation/sklearn.py (hunk 9 above)
Comment (mingkang111) [reply guardian_ai/fairness/bias_mitigation/sklearn.py]: After discussing offline, leveling down will be >= 0. Made adjustment on `show_tradeoff`.

File: docs/source/user_guide/fairness/quickstart.rst [hunk 10]
```diff
@@ -104,6 +104,10 @@ Bias Mitigation
         if col in numeric_columns:
             df[col] = df[col].astype(int)
 
+    X_train, X_test, y_train, y_test = train_test_split(
```
Comment (YashaPushak) [inline docs/source/user_guide/fairness/quickstart.rst:107]: Did the old code work without mapping to 0 and 1? I thought that was already a requirement for this module.

File: guardian_ai/fairness/bias_mitigation/sklearn.py [hunk 11]
```diff
@@ -229,6 +246,7 @@ def __init__(
         regularization_factor: float = 1e-3,
         favorable_label_idx: int = 1,
         random_seed: int = 0,
+        third_objective: bool = True,
```
Comment (YashaPushak) [inline guardian_ai/fairness/bias_mitigation/sklearn.py:249]: The user doesn't really need this configuration option. I think it was mostly added for internal testing purposes. I recommend un-exposing it, although it doesn't hurt to keep it as a private attribute that is set to `True` by default.

File: guardian_ai/fairness/bias_mitigation/sklearn.py [hunk 12]
```diff
@@ -1141,31 +1363,73 @@ def show_tradeoff(self, hide_inadmissible: bool = False):
...
                 text=df["index"],
                 line_shape="vh" if self._higher_fairness_is_better else "hv",
-                mode="markers+lines",
+                mode="markers" if metric_is_valid else "markers+lines",
+                marker=marker,
                 hovertemplate=f"{self.fairness_metric_name}"
                 + ": %{x:.4f}"
                 + f"<br>{self.accuracy_metric_name}"
                 + ": %{y:.4f}</br>"
+                + f"<br>{TPR_or_level_down}"
```
Comment (YashaPushak) [inline guardian_ai/fairness/bias_mitigation/sklearn.py:1410]: - [ ] What does this show up as to the user?

File: docs/source/user_guide/fairness/quickstart.rst (hunk 10 above)
Comment (mingkang111) [reply docs/source/user_guide/fairness/quickstart.rst:107]: The old code doesn't work without this line.

File: guardian_ai/fairness/bias_mitigation/sklearn.py (hunk 12 above)
Comment (mingkang111) [reply guardian_ai/fairness/bias_mitigation/sklearn.py:1410]: here is how it determine: https://github.com/oracle/guardian-ai/blob/3ad20a32ed9fd62b2f5465b963a878938d444f31/guardian_ai/fairness/bias_mitigation/sklearn.py#L1367

The old code is showing `"TPR Difference"`.

Comment (YashaPushak) [reply guardian_ai/fairness/bias_mitigation/sklearn.py:1410]: Aah, okay. On line 1367 you can put:
`level_down = "</br>Levelling Down: %{customdata:.4f}</br>" if metric_is_valid else ""`
and then here you can display that instead. If the metric does not support calculating levelling down then we should simply not print anything here about it.

Comment (mingkang111) [reply guardian_ai/fairness/bias_mitigation/sklearn.py]: Thanks @YashaPushak ! Fixed.

File: guardian_ai/fairness/bias_mitigation/sklearn.py (hunk 11 above)
Comment (mingkang111) [reply guardian_ai/fairness/bias_mitigation/sklearn.py]: Removed it from init and docstring.

File: guardian_ai/fairness/bias_mitigation/sklearn.py [hunk 13]
```diff
@@ -1141,31 +1369,71 @@ def show_tradeoff(self, hide_inadmissible: bool = False):
             Whether or not to hide the models that don't satisfy the
             constraint.
         """
+        metric_is_valid = self.fairness_metric_name in _valid_regression_metrics
+        level_down = "levelling_down: %{customdata:.4f}" if metric_is_valid else ""
+        index_ = "<br>Index: %{text}</br>" if metric_is_valid else "Index: %{text}"
```
Comment (YashaPushak) [inline guardian_ai/fairness/bias_mitigation/sklearn.py:1374]: Should the break before index only be there if level_down is not an empty string? 

Maybe there should just always be a break at the end of each line, so that we don't need to check if a line is empty to decide to insert a line at the beginning of the next line.

**Review by YashaPushak — APPROVED** [review]
Approving now because my last comment is minor and I'm confident that you can resolve it without needing me to re-review.

File: guardian_ai/fairness/bias_mitigation/sklearn.py (hunk 13 above)
Comment (mingkang111) [reply guardian_ai/fairness/bias_mitigation/sklearn.py:1374]: Since level_down doesn't have break (aims to avoid extra blank line between `accuracy_metric_name` and level_down),  it is necessary to introduce a break for the index. This ensures that the index is not appended directly to the end of the level_down line.

File: guardian_ai/fairness/bias_mitigation/sklearn.py [hunk 14]
```diff
@@ -1408,12 +1409,12 @@ def show_tradeoff(self, hide_inadmissible: bool = False):
                 line_shape="vh" if self._higher_fairness_is_better else "hv",
                 mode="markers" if metric_is_valid else "markers+lines",
                 marker=marker,
-                hovertemplate=f"{self.fairness_metric_name}"
-                + ": %{x:.4f}"
+                hovertemplate=f"<br>{self.fairness_metric_name}"
```
Comment (YashaPushak) [inline guardian_ai/fairness/bias_mitigation/sklearn.py:1412]: It looks like this will give two line breaks for every line -- i.e., there will be a blank line between every item. I guess that's fine (it's consistent), but I was actually suggesting that you remove the <br> tag that is at the start of every line so that there is never any empty lines.

Comment (mingkang111) [reply guardian_ai/fairness/bias_mitigation/sklearn.py]: Thanks for clarifying! I have updated the change.

**Review by YashaPushak — APPROVED** [review]
Thanks Ming!
